File: client/src/api/protocol.ts
export type SessionRole = "alice" | "bob";

export type ClientEvent =
  | { type: "join_session"; session_id: string; role: SessionRole }
  | { type: "handshake_offer"; offer_x25519_public: string; alice_nonce: string }
  | {
      type: "handshake_answer";
      bob_mlkem_public: string;
      bob_x25519_public: string;
      bob_nonce: string;
    }
  | { type: "handshake_finish"; kem_ciphertext: string; mac: string }
  | { type: "handshake_confirm"; mac: string }
  | { type: "chat_message"; seq: number; nonce: string; ciphertext: string }
  | { type: "chat_ack"; seq: number }
  | {
      type: "file_offer";
      transfer_id: string;
      name: string;
      mime_type: string;
      size: number;
      total_chunks: number;
      sha256: string;
    }
  | { type: "file_accept"; transfer_id: string }
  | { type: "file_reject"; transfer_id: string }
  | { type: "file_complete"; transfer_id: string }
  | { type: "file_abort"; transfer_id: string; reason: string }
  | { type: "file_resume_state"; transfer_id: string; received_bitmap: string }
  | { type: "register_resume"; verifier: string }
  | { type: "resume_session"; session_id: string; role: SessionRole }
  | { type: "resume_proof"; resume_key: string; mac: string }
  | { type: "leave_session" }
  | { type: "ping" };

export type ServerEvent =
  | { type: "joined_session"; role: SessionRole; expires_in_seconds: number }
  | { type: "peer_joined" }
  | { type: "relay_handshake_offer"; offer_x25519_public: string; alice_nonce: string }
  | {
      type: "relay_handshake_answer";
      bob_mlkem_public: string;
      bob_x25519_public: string;
      bob_nonce: string;
    }
  | { type: "relay_handshake_finish"; kem_ciphertext: string; mac: string }
  | { type: "relay_handshake_confirm"; mac: string }
  | { type: "relay_chat_message"; seq: number; nonce: string; ciphertext: string }
  | { type: "relay_chat_ack"; seq: number }
  | {
      type: "relay_file_offer";
      transfer_id: string;
      name: string;
      mime_type: string;
      size: number;
      total_chunks: number;
      sha256: string;
    }
  | { type: "relay_file_accept"; transfer_id: string }
  | { type: "relay_file_reject"; transfer_id: string }
  | { type: "relay_file_complete"; transfer_id: string }
  | { type: "relay_file_abort"; transfer_id: string; reason: string }
  | { type: "relay_file_resume_state"; transfer_id: string; received_bitmap: string }
  | { type: "resume_challenge"; nonce: string }
  | { type: "resume_accepted" }
  | { type: "peer_reconnected" }
  | { type: "peer_left" }
  | { type: "session_expired" }
  | { type: "pong" }
  | { type: "error"; message: string };

File: server/src/ws/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum SessionRole {
    Alice,
    Bob,
}

#[derive(Debug, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientEvent {
    JoinSession {
        session_id: String,
        role: SessionRole,
    },
    HandshakeOffer {
        offer_x25519_public: String,
        alice_nonce: String,
    },
    HandshakeAnswer {
        bob_mlkem_public: String,
        bob_x25519_public: String,
        bob_nonce: String,
    },
    HandshakeFinish {
        kem_ciphertext: String,
        mac: String,
    },
    HandshakeConfirm {
        mac: String,
    },
    ChatMessage {
        seq: u64,
        nonce: String,
        ciphertext: String,
    },
    ChatAck {
        seq: u64,
    },
    FileOffer {
        transfer_id: String,
        name: String,
        mime_type: String,
        size: u64,
        total_chunks: u32,
        sha256: String,
    },
    FileAccept {
        transfer_id: String,
    },
    FileReject {
        transfer_id: String,
    },
    FileComplete {
        transfer_id: String,
    },
    FileAbort {
        transfer_id: String,
        reason: String,
    },
    FileResumeState {
        transfer_id: String,
        received_bitmap: String,
    },
    RegisterResume {
        verifier: String,
    },
    ResumeSession {
        session_id: String,
        role: SessionRole,
    },
    ResumeProof {
        resume_key: String,
        mac: String,
    },
    LeaveSession,
    Ping,
}

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerEvent {
    JoinedSession {
        role: SessionRole,
        expires_in_seconds: u64,
    },
    PeerJoined,
    RelayHandshakeOffer {
        offer_x25519_public: String,
        alice_nonce: String,
    },
    RelayHandshakeAnswer {
        bob_mlkem_public: String,
        bob_x25519_public: String,
        bob_nonce: String,
    },
    RelayHandshakeFinish {
        kem_ciphertext: String,
        mac: String,
    },
    RelayHandshakeConfirm {
        mac: String,
    },
    RelayChatMessage {
        seq: u64,
        nonce: String,
        ciphertext: String,
    },
    RelayChatAck {
        seq: u64,
    },
    RelayFileOffer {
        transfer_id: String,
        name: String,
        mime_type: String,
        size: u64,
        total_chunks: u32,
        sha256: String,
    },
    RelayFileAccept {
        transfer_id: String,
    },
    RelayFileReject {
        transfer_id: String,
    },
    RelayFileComplete {
        transfer_id: String,
    },
    RelayFileAbort {
        transfer_id: String,
        reason: String,
    },
    RelayFileResumeState {
        transfer_id: String,
        received_bitmap: String,
    },
    ResumeChallenge {
        nonce: String,
    },
    ResumeAccepted,
    PeerReconnected,
    PeerLeft,
    SessionExpired,
    Error {
        message: String,
    },
    Pong,
}

#[derive(Debug, Clone)]
pub enum RelayPayload {
    Event(ServerEvent),
    Binary(Vec<u8>),
}
